### ISSUE TITLE
Proposal for API changes using latest Microstates

### ISSUE THREAD

**Issue by taras** [microstates/ember#70]
Related: #68 & #69

There has been a lot of change to Microstates since the last version of this addon was released. This issue is a proposal for changes to give us an opportunity to discuss before making the changes. 

I've been experimenting with different ways of using Microstates in Ember and React. I keep returning to a pattern when components have a Microstate that represents it's state. This makes it very easy to bind microstates to Handlebars templates. There is very little ceremony in setting up a Microstate when using `ember-auto-import`.

## In the component class

```js
import Component from '@ember/component';
import hbs from 'htmlbars-inline-precompile';
import { create, Store } from 'microstates';

class Person {
  firstName = String;
  lastName = String;
}

export default Component.extend({
  layout: hbs`{{$.firstName.state}} {{$.lastName.state}}`,
  init() {
    let firstName = this.get('firstName');
    let lastName = this.get('lastName');

    let state = create(Person, { firstName, lastName });
    this.set('$', Store(state, next => this.set('$', next)));
  }
});
```

With that said, we might want to abstract this setup into a macro. So, it would look like this.

```js
import Component from '@ember/component';
import hbs from 'htmlbars-inline-precompile';
import { useState } from 'ember-microstates';
import { create } from 'microstates';

class Person {
  firstName = String;
  lastName = String;
}

export default Component.extend({
  layout: hbs`{{$.firstName.state}} {{$.lastName.state}}`,
  $: useState(create(Person)),

  didReceiveAttrs() {
    this._super(...arguments);

    let firstName = this.get('firstName');
    let lastName = this.get('lastName');

    this.get('$')
      .firstName.set(firstName)
      .lastName.set(lastName);
  }
});
```

*Note*: It should be safe to call `set` in didReceiveAttrs because changes that do not change the value are debounced as no-op by Microstates.  

## In the template

I would like to be able to setup Microstates in the template without changing the component JavaScript class. For this, we would have a `use-state` helper. This helper would take a microstate and wrap it in a store. It would be used with the `let` help to make the microstate available in the context. 

One possibly contentious topic is what arguments `use-state` should accept. I believe that `use-state` should accept a Microstate instance and not a type because creating a Microstate in the template without using JavaScript is going to be problematic. There are several challenges that make this difficult: 

1. There is no way to import arbitrary modules into the template. If there was an `import` helper then you could import a type from a module and create a Microstate. Without `import` you have to make sure that the type is in the context via JavaScript before passing it to `use-state`.

2. Mapping properties to microstates can be problematic. If you create a microstate in the template, you need to give it initial value. That value needs to map to the shape of the microstate. Doing this in the template can get very deeply nested. If that is necessary, then it might better to be done in JavaScript.

3. Creating microstates in handlebars will cause a new microstate to be created every time that the helper is re-evaluated. Microstates relies heavily on structural reuse. It will be more practical and safer for Microstates to not get values applied automatically and instead for them to be applied manually when necessary.

For these reasons, I think `use-state` should expect to receive a microstate instance or a string that can be used to lookup a microstate. To make the lookup possible, we'll register a new type called `state`. Microstates will be stored in `states` directory. The directory will contain files that have a microstate instance as a default export. Using the `states` directory will be optional because any helper that resolves a microstate from a string, can also take a microstate. 

Here is what using `use-state` helper would look like:

```hbs
{{#let (use-state 'person') as |person|}}
  <p>{{person.name.state}}</p>
{{/let}}
```

The above example has very limited use because it only uses original microstates' content. A real use case would include capturing user's input. For this we need to introduce 2nd argument which will send the original microstate that is the result of the transition.

```hbs
{{#let (use-state (or last 'person') (action (mut last))) as |person|}}
  <p>{{person.name.state}}</p>
{{/let}}
```

This pattern would give access to the state of the microstate in both JavaScript and the template. It's worth nothing that I think the callback should receive the source microstate not the identity.

What to do you think?

**Comment by cowboyd** [comment]
I think we want the solution to scale from individual template use, to something shared by multiple components, or even different pieces of the application.

1. You should be able to use a microstate from only within a template and never having to resort to javascript. I'm actually an advocate pushing static imports, in the same way that this project was the impetus for `let`. We can also get away with a less fully generic `let` helper.

```hbs
{{type Box="./states/box"}}
{{let box=(Box height=500 width=200)}}
{{box.height.state}} X {{box.width.state}}
area: {{box.area}}
```

2. You should be able to instantiate it from within a JavaScript component and make it available to the template.

```js
import { useState } from "@microstates/ember";
import Box from "./states/box";

@useState(Box, "box", {height: 100, width: 200});
export default class extends Component {
  constructor() {
    console.log('the box is', this.get('box');
  }
}
```

By default this thing will get a 100x200 box, but you can also pass it in as the `box` property:

```hbs
<MyComponent @box={{box-from-elsewhere}}/>
```
It can type-check to make sure that the box is indeed an instance of box.

**Comment by taras** [comment]
`{{type Box="./states/box"}}` is not a valid syntax for this case. It'd need to use `let` because Ember only has block style let helper.

```hbs
{{#let (type "./states/box") as | Box|}}
  ...
{{/let}}
```

`{{let box=(Box height=500 width=200)}}` is not a valid syntax because you can't use a variable as a helper. 

it would need to be 

```hbs
{{#let (type "./states/box") as | Box|}}
  {{#let (create Box height=500 width=500) as |box|}}
    {{box.height.state}} X {{box.width.state}}
     area: {{box.area}}
  {{/let}}
{{/let}}
```

It gets quiet nested, so we can flatten it and rename type to import.

```hbs
  {{#let (create (import "./states/box")  height=500 width=500) as |box|}}
    {{box.height.state}} X {{box.width.state}}
     area: {{box.area}}
  {{/let}}
```

I'm not a big fan of `useState` decorator because it introduces the middle argument. `useState` computed property would be less unusual in ember world.

```js
import Component from '@ember/component';
import { useState } from "@microstates/ember";
import Box from "./states/box";

export default Component.extend({
  box: useState(Box,  {height: 100, width: 200})
  constructor() {
    console.log('the box is', this.get('box');
  }
})
```

To make `use-state` consistent, we can call create helper `use-state`.

```hbs
  {{#let (use-state (import "./states/box")  height=500 width=500) as |box|}}
    {{box.height.state}} X {{box.width.state}}
     area: {{box.area}}
  {{/let}}
```

**Comment by knownasilya** [comment]
I'd be up for a decorator and a CP approach, both are used depending if people are using decorators or not.

**Comment by taras** [comment]
I would prefer to not do the decorator because I don't like the property where the microstate will be mounted. 

@knownasilya what do you think about the `use-state` or `use-type` helper?

**Comment by knownasilya** [comment]
Regarding decorator I'd do:

```js
@useState(Box, {height: 100, width: 200})
box;
```

But yes, a CP should come first.
Regarding the helpers, I like `use-type`.

**Comment by donaldwasserman** [comment]
@taras @knownasilya - I've been using/experimenting with microstates.js on a project and really like it. (I just came across this addon when looking up something in the microstates docs). As a user, I really like the helper approach (just using the `use-state` name for now).

I like the sytanx that @taras outlined, but rather than registering all the states and look them up, would be it be better to use a helper to return a states instance? 
```hbs
{{#let (use-state (ms-person (hash firstName="jenny" lastName="smith"))) as |person|}}
  <p>{{person.name.state}}</p>
{{/let}}
```
It's a little more verbose, but seems simplier to implement, and allows per-use specific options, ie: 

`use-state (ms-string "hello")` vs `use-state (ms-array (array foo, bar, baz))`

I'd love to help on this any way I can, I think this is a great idea.

**Comment by taras** [comment]
As a user, I really like the helper approach (just using the use-state name for now).

I'm leaning towards using `useType` and `use-type` to be symmetrical with `useType` hook in `@microstates/react`. 

> I like the sytanx that @taras outlined, but rather than registering all the states and look them up, would be it be better to use a helper to return a states instance?

One of the challenges with how Ember registers helpers is that there is no easy way to define sub-expressions like `(ms-person)`. We have to explicitly register them. We can do what Ember does for helpers and register helpers for types. 

For example, you can create `person` type and it'll automatically register `(person )` helper. Params that are sent to this helper will be passed to the create function. 

Then we have essentially 2 APIs. 

1. `(use-type (import 'my-app/someplace/type') value)` which imports a type from any path and creates a microstate
2. `(my-type value)` which uses the implicitly registered `my-type` helper and instantiates that microstate.

This should pretty comfortable and really super power the templates.

**Comment by taras** [comment]
BTW, I would love help on this because I don’t use Ember often so I don’t have a lot of motivation or even remember much other than by mussle memory.

**Comment by taras** [comment]
> So in the use-type helper, gets a string or microstate instance and an action.

No, `use-type` will accept Type and value. For example, `(use-type Number 10)`.

> The only thing i don't understand is the Store where does that come from?

It's not documented correctly. I explained it here. We need to add it to the README

**Comment by taras** [comment]
Resolved by #72
